conformance/conformance_api.py

```diff
@@ -0,0 +1,1319 @@
+* The actual request is not an exact match of the request specified by the test data.
+* The server fails to create a response or error that is an exact match of the specified
+  response or error.
+
+When checking for an exact match, the only thing that can be ignored is the order of the
+properties of an object. There should be no missing properties, no extra properties, and
+all property names and other strings should use the same letter case.
+
+## Conformance Tester
+
+A conformance tester should leverage a Facility client generator and/or library. The tester
+should enumerate through all of the conformance tests. For each test, it should create a
+client that adds the `FacilityTest` HTTP header to any request. It should use that client to
+call the method specified by the test data with the request specified by the test data. The
+tester should verify that the result matches the response or error specified by the test data.
+
+The tester should report a test failure if:
+
+* The method specified by the test data is missing.
+* The client fails to create a request that is an exact match of the specified request.
+* The client receives a response or error that is not an exact match of the specified
+  response or error.
+"""
+
+import base64
+import decimal
+import typing
+
+import facility
+
+
+HTTP_STATUS_CODE_TO_ERROR_CODE = {sc: ec for ec, sc in facility.ERROR_CODE_TO_HTTP_STATUS_CODE.items()}
+HTTP_STATUS_CODE_TO_ERROR_CODE[403] = "Forbidden"  # ApiErrors.NotAdmin: The user is not an administrator.
+HTTP_STATUS_CODE_TO_ERROR_CODE[500] = "InternalServerError"  # ApiErrors.TooHappy: I'm "too" 😄!
+
+
+class Answer(facility.Enum):
+    """
+    One of three answers.
+
+    Use `maybe` if you aren't sure.
+    """
+    YES = "yes"  # Affirmative.
+    NO = "no"  # Negative.
+    MAYBE = "maybe"  # Unknown.
+
+
+class Widget(facility.DTO):
+    """
+    A widget.
+
+    Widgets are useful for conformance tests.
+    """
+    def __init__(
+        self,
+        id_: int,
+        name: str,
+    ):
+        """
+        :param id_: A unique identifier for the widget.
+        :param name: The name of the widget.
+        """
+        super().__init__()
+        if not isinstance(id_, (int, type(None))):
+            raise ValueError(f"Invalid id_: {id_}")
+        self.id_ = id_
+        if not isinstance(name, (str, type(None))):
+            raise ValueError(f"Invalid name: {name}")
+        self.name = name
+
+    @classmethod
+    def from_data(cls, data: typing.Dict[str, typing.Any]) -> "Widget":
+        return Widget(
+            id_=int(data["id"]) if "id" in data else None,
+            name=data.get("name"),
+        )
+
+
+class Any(facility.DTO):
+    def __init__(
+        self,
+        string: str,
+        boolean: bool,
+        double: float,
+        int32: int,
+        int64: int,
+        decimal_: decimal.Decimal,
+        bytes_: bytes,
+        object_: object,
+        error: facility.Error,
+        data: "Any",
```

Comment:
I'm surprised `Any` is in quotes here. Python type names can be quoted? Or am I misunderstanding what this means?

Reply:
The Python type checker won't understand it without quotes until after the class definition, because it's not defined yet. (I'm not sure why; I think it should.) But it works with quotes.